src/eepromfs.c

```diff
@@ -0,0 +1,888 @@
+ * and deallocated when the filesystem is de-initialized.
+ */
+static eepfs_file_t *eepfs_files = NULL;
+
+/**
+ * @brief The filesystem prefix used by stdio functions.
+ * 
+ * This is an optional parameter when initializing the
+ * filesystem, but allows for `fopen` and friends to access
+ * the EEPROM instead of using `eepfs_*` methods.
+ * 
+ * This prefix is captured so that it can be detached when
+ * the filesystem de-initializes.
+ */
+static const char * eepfs_stdio_prefix = NULL;
+
+/**
+ * @brief Finds a file descriptor's handle given a path.
+ * 
+ * A leading '/' in the path is optional and is supported to
+ * allow for consistency with stdio functions.
+ *
+ * @param[in] path
+ *            Absolute path of the file to open (without stdio prefix)
+ *
+ * @return A file handle or a negative value on failure.
+ */
+static int eepfs_find_handle(const char * const path)
+{
+    /* Sanity check */
+    if ( path == NULL )
+    {
+        return EEPFS_EBADINPUT;
+    }
+
+    /* Ignore the leading '/' on the path (if present) */
+    const char * normalized_path = path + (path[0] == '/');
+
+    /* Check if the path matches an entry in the #eepfs_files array */
+    for ( size_t i = 1; i < eepfs_num_files; ++i )
+    {
+        if ( strcmp(normalized_path, eepfs_files[i].path) == 0 )
+        {
+            /* A file exists matching that path */
+            return i;
+        }
+    }
+
+    /* File does not exist */
+    return EEPFS_ENOFILE;
+}
+
+/**
+ * @brief Resolves a file handle to a file descriptor.
+ * 
+ * File descriptors can be used for reading or writing.
+ * 
+ * Getting a file repeatedly will always return the same pointer.
+ * This means that the cursor is shared across all usage of the file.
+ * Be sure to reset the cursor when you are done with it!
+ * 
+ * Negative handles indicate an upstream error occurred.
+ * The zero value will never be used as a file handle, even
+ * though it does map to an entry in the `eepfs_files` table.
+ * The signature entry is not intended to be used as a file.
+ * 
+ * @param[in] handle
+ *            The file handle of an entry in the `eepfs_files` table
+ * 
+ * @return A pointer to the file descriptor or NULL if the handle is invalid
+ */
+static eepfs_file_t * const eepfs_get_file(const int handle)
+{
+    /* Check if the handle appears to be valid */
+    if ( handle > 0 && handle < eepfs_num_files )
+    {
+        return &eepfs_files[handle];
+    }
+
+    return NULL;
+}
+
+/**
+ * @brief Reads a specific amount of data from a file starting from the cursor.
+ *
+ * @param[in]  path
+ *             Path of file in EEPROM filesystem to read from
+ * @param[out] dest
+ *             Buffer to read into
+ * @param[in]  len
+ *             Number of bytes to read
+ *
+ * @return The actual number of bytes read or a negative value on failure.
+ */
+static int eepfs_cursor_read(eepfs_file_t * const file, void * const dest, const size_t len)
```

Comment:
Now that I think of it: maybe `fopen` in newlib already provides a buffering layer? I'm not sure how that interacts with eepfs and I'm sure it's not designed around the concept of underlying fixed size blocks, but maybe it's something to take into account.

Reply:
I checked newlib sources and it looks like newlib by default adds a 1Kb buffer to all stdio files. I'm not sure on how that interacts with seeking for instance, but for simple cases it looks like this is useful for us because it means that we would probably get a single read/write call for the whole file as soon as the buffer is flushed (at fclose time, or in case of explicit fflush -- and then I don't know if seeking also causes a flush). This is something I'd love to be investigated and documented though.